server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -77,47 +77,43 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
         final List<FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedList(new ArrayList<>());
-        final ActionListener<List<FieldCapabilitiesIndexResponse>> countDownListener = new ActionListener<>() {
-            @Override
-            public void onResponse(List<FieldCapabilitiesIndexResponse> results) {
-                for (FieldCapabilitiesIndexResponse res : results) {
-                    if (res.canMatch()) {
-                        indexResponses.add(res);
-                    }
-                }
-                countDown();
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                // TODO we should somehow inform the user that we failed
-                countDown();
-            }
 
-            private void countDown() {
-                if (completionCounter.countDown()) {
-                    if (request.isMergeResults()) {
-                        listener.onResponse(merge(indexResponses, request.includeUnmapped()));
-                    } else {
-                        listener.onResponse(new FieldCapabilitiesResponse(indexResponses));
-                    }
+        final Runnable countDown = () -> {
+            if (completionCounter.countDown()) {
+                if (request.isMergeResults()) {
+                    listener.onResponse(merge(indexResponses, request.includeUnmapped()));
+                } else {
+                    listener.onResponse(new FieldCapabilitiesResponse(indexResponses));
                 }
             }
         };
 
         for (String index : concreteIndices) {
-            client.executeLocally(TransportFieldCapabilitiesIndexAction.TYPE,
+            client.executeLocally(
+                TransportFieldCapabilitiesIndexAction.TYPE,
                 new FieldCapabilitiesIndexRequest(
                     request.fields(),
                     index,
                     localIndices,
                     request.indexFilter(),
-                    nowInMillis, request.runtimeFields()
+                    nowInMillis,
+                    request.runtimeFields()
                 ),
-                ActionListener.wrap(
-                    response -> countDownListener.onResponse(Collections.singletonList(response)),
-                    countDownListener::onFailure
-                )
+                new ActionListener<FieldCapabilitiesIndexResponse>() {
+                    @Override
+                    public void onResponse(FieldCapabilitiesIndexResponse result) {
+                        if (result.canMatch()) {
+                            indexResponses.add(result);
+                        }
+                        countDown.run();
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        indexResponses.add(new FieldCapabilitiesIndexResponse(index, e));
```

Comment:
We can gather the exception in a synchronized map, no need to build fake `FieldCapabilitiesIndexResponse`.